arch/x86/adler32_avx512.c

```diff
@@ -0,0 +1,113 @@
+/* adler32_avx512.c -- compute the Adler-32 checksum of a data stream
+ * Copyright (C) 1995-2011 Mark Adler
+ * Authors:
+ *   Brian Bockelman <bockelman@gmail.com>
+ *   Adam Stylinski <kungfujesus06@gmail.com>
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "../../zbuild.h"
+#include "../../zutil.h"
+
+#include "../../adler32_p.h"
+
+#include <immintrin.h>
+#include <x86intrin.h>
+
+#ifdef X86_AVX512_ADLER32
+
+Z_INTERNAL uint32_t adler32_avx512(uint32_t adler, const unsigned char *buf, size_t len) {
+    uint32_t sum2;
+
+     /* split Adler-32 into component sums */
+    sum2 = (adler >> 16) & 0xffff;
+    adler &= 0xffff;
+
+    /* in case user likes doing a byte at a time, keep it fast */
+    if (UNLIKELY(len == 1))
+        return adler32_len_1(adler, buf, sum2);
+
+    /* initial Adler-32 value (deferred check for len == 1 speed) */
+    if (UNLIKELY(buf == NULL))
+        return 1L;
+
+    /* in case short lengths are provided, keep it somewhat fast */
+    if (UNLIKELY(len < 16))
+        return adler32_len_16(adler, buf, len, sum2);
+
+    const __mmask16 vs_mask = 1U << 15;
+    __m512i vs1 = _mm512_maskz_set1_epi32(vs_mask, adler);
+    __m512i vs2 = _mm512_maskz_set1_epi32(vs_mask, sum2);
+
+    const __m512i dot1v = _mm512_set1_epi8(1);
+    /* No idea why there's no setr but I guess we'll deal with it */
+    const __m512i dot2v = _mm512_set_epi8(1, 2, 3, 4, 5, 6,
+                                          7, 8, 9, 10, 11, 12, 13,
```

Comment:
Maybe condense this matrix into 3-4 lines unless there is a good reason?

Reply:
None other than my 80 column bias